pkg/sensors/exec/cgroups_test.go

```diff
@@ -1068,3 +1097,247 @@ func TestCgroupv1K8sHierarchyInHybridBlkio(t *testing.T) {
 func TestCgroupv1K8sHierarchyInHybridInvalid(t *testing.T) {
 	testCgroupv1K8sHierarchyInHybrid(t, "invalid-cgroup-controller")
 }
+
+// This will execute a helper program that creates cgroups
+// then migrates to them and perform for each one exec in order
+// to assert collected cgroup id + name of the exec event match.
+// This works in Hybrid mode and is performed for all hierarchies.
+func testCgroupv1ExecK8sHierarchyInHybrid(t *testing.T, selectedController string) {
```

Comment:
It seems like a lot of the code here has been copy-pasted from `testCgroupv1K8sHierarchyInHybrid`.
Please add helper functions to i) avoid code duplication and ii) make it clearer what the code actual does.

Reply:
We can but for the time being it is not easy, those tests are different

- `testCgroupv1K8sHierarchyInHybrid` does test only mkdir / rmdir events probing other cgroup controllers to ensure that we catch only events that belong to the right controller/hierarchy we are interested in.

- `testCgroupv1ExecK8sHierarchyInHybrid` does the mkdir / execve / rmdir but only on the pids / memory controllers without probing other controllers as we should not see their events (this is already handled by the above test). This test handles only this case of the execve, but also there is a bug in our CI for old 4.9 kernels where it fails when execve migrating on the cpuset controller, where local tests on same kernel succeeded... don't know what's the cause... 

I think the code that could be shared is already shared, but I will try to see if I can share more, I think I will add more comments to explain why and trying to keep things clear even if we have some lines duplicated, changing one test won't affect the other.

Reply:
Given that this is the diff between the two functions:

```diff

-func testCgroupv1K8sHierarchyInHybrid(t *testing.T, selectedController string) {
+func testCgroupv1ExecK8sHierarchyInHybrid(t *testing.T, selectedController string) {
 	testutils.CaptureLog(t, logger.GetLogger().(*logrus.Logger))
 	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
 	defer cancel()
@@ -69,9 +68,9 @@
 		}
 
 		/* If not found falls back to default selected controller by
-		 * the cgroups auto discovery.
+		 * the cgroups auto discovery. For now we only do memory and pids controllers
 		 */
-		if !found {
+		if !found || (usedController != "memory" && usedController != "pids") {
 			usedController = cgroups.GetCgrpControllerName()
 		}
 	}
@@ -82,6 +81,8 @@
 	controllers := append([]cgroupController(nil), cgroupv1Controllers...)
 	kubeCgroupHierarchiesMap := make(map[string][]cgroupHierarchy, len(cgroupv1Controllers))
 
+	// Register cleanups first
+
 	// Ensure we unmount all cgroup mount points
 	cleanupMounts := func() {
 		for _, controller := range controllers {
@@ -129,8 +130,12 @@
 				} else {
 					n.path = c.path
 				}
-				if controller.name == usedController && uint32(i) < trackingCgrpLevel {
-					n.tracking = true
+				if controller.name == usedController {
+					if uint32(i) < trackingCgrpLevel {
+						n.tracking = true
+					}
+					// We are tracking this controller expect docker fields
+					n.expectedDocker = c.expectedDocker
 				} else {
 					n.tracking = false // Controller was not used at all, hierarchy not to be tracked
 				}
@@ -147,45 +152,72 @@
 	err = logCgroupMountInfo(t)
 	assert.NoError(t, err)
 
-	triggerCgroupMkdir := func() {
-		for hierarchy, cgroupHierarchy := range kubeCgroupHierarchiesMap {
-			path := ""
-			for _, dir := range cgroupHierarchy {
-				path = filepath.Join(path, dir.path)
-				err = cgroupMkdir(t, cgroupRoot, hierarchy, path)
-				if err != nil {
-					t.Fatalf("Failed to create cgroup %s/%s/%s", cgroupRoot, hierarchy, dir.path)
-				}
+	testCgroupMigrate := testutils.ContribPath("tester-progs/cgroup-migrate.bash")
+	triggerCgroupExec := func() {
+		path := filepath.Join(cgroupRoot, usedController)
+		for i, dir := range kubeCgroupHierarchiesMap[usedController] {
+			path = filepath.Join(path, dir.path)
+			if dir.tracking == true {
+				kubeCgroupHierarchiesMap[usedController][i].matchExecCgrpID = true
+			}
+			/* We expect docker field here */
+			if dir.expectedDocker == "yes" {
+				docker, _ := procevents.LookupContainerId(dir.path, true, false)
+				require.NotEmpty(t, docker)
+				kubeCgroupHierarchiesMap[usedController][i].expectedDocker = docker
+			}
+			var outb, errb bytes.Buffer
+			cmd := exec.Command(testCgroupMigrate, "--mode", "cgroupv1", "--controller", usedController,
+				"--new", dir.path, path)
+			cmd.Stdout = &outb
+			cmd.Stderr = &errb
+			t.Logf("Executing command: %v", cmd.String())
+			err := cmd.Run()
+			stderr := errb.String()
+			if len(stderr) > 0 {
+				t.Logf("\nstderr:\n%v", stderr)
+			}
+			if err != nil {
+				t.Fatalf("Command failed: %s", err)
+			}
+			if len(outb.String()) > 0 {
+				t.Logf("\nstdout:\n%v\n", outb.String())
 			}
 		}
 	}
 
 	triggerCgroupRmdir := func() {
-		for hierarchy, cgroupHierarchy := range kubeCgroupHierarchiesMap {
-			path := "/"
-			for _, dir := range cgroupHierarchy {
-				path = filepath.Join(path, dir.path)
-			}
-
-			for i := 0; i < len(cgroupHierarchy); i++ {
-				err := cgroupRmdir(t, cgroupRoot, hierarchy, path)
-				if err != nil {
-					t.Fatalf("Failed to remove cgroup %s/%s/%s", cgroupRoot, hierarchy, path)
-				}
-				path = filepath.Dir(path)
+		path := "/"
+		for _, dir := range kubeCgroupHierarchiesMap[usedController] {
+			path = filepath.Join(path, dir.path)
+		}
+		for i := 0; i < len(kubeCgroupHierarchiesMap[usedController]); i++ {
+			err := cgroupRmdir(t, cgroupRoot, usedController, path)
+			if err != nil {
+				t.Fatalf("Failed to remove cgroup %s/%s/%s", cgroupRoot, usedController, path)
 			}
+			path = filepath.Dir(path)
 		}
 	}
 
 	triggers := []func(){
-		triggerCgroupMkdir, triggerCgroupRmdir,
+		triggerCgroupExec, triggerCgroupRmdir,
 	}
 
 	// Loop over all created cgroup hierarchies
 	testCgroupv1HierarchyInHybrid(ctx, t, cgroupRoot, usedController, kubeCgroupHierarchiesMap,
 		trackingCgrpLevel, triggers)
 
+	// dump final constructed values
+	for controller, cgroupHierarchy := range kubeCgroupHierarchiesMap {
+		t.Logf("\nkubeCgroupHierarchiesMap[%s]=%+v\n", controller, cgroupHierarchy)
+	}
+
 	for _, cgroupHierarchy := range kubeCgroupHierarchiesMap {
 		assertCgroupDirTracking(t, cgroupHierarchy)
 	}
+
+	for _, cgroupHierarchy := range kubeCgroupHierarchiesMap {
+		assertCgroupIDsTracking(t, cgroupHierarchy)
+	}
 }
```

it seems to me that the main difference is the triggers, and those can be passed as arguments.

> I think I will add more comments to explain why and trying to keep things clear even if we have some lines duplicated, changing one test won't affect the other.

I think that also makes sense. The code does not seem easy to follow. I'd rather have simpler tests, but if we cannot we should at least make them so as understandable as possible both for other people but also for our future selves.

Reply:
> it seems to me that the main difference is the triggers, and those can be passed as arguments.

You are missing the controllers..., please take a look again, this diff shows exactly why they differ, without exec is trying events on all controller but ensures that we only work on one controller or hierarchy mkdir/rmdir events and again without exec context at all. Where the new one with exec just works on the desired controller we don't use other controllers as there is no point to test them...

That test is written explicitly like this to show that the `testCgroupv1ExecK8sHierarchyInHybrid` has the exec context hence in the name of test and we test it only desired controllers/hierarchies due the nature of the execve event and how things work in production... that's why the test are written like this... mixing them or sharing code will just add confusions

Reply:
Ok, added comments to tests to make improve readability.

Reply:
It's still seems to me that there is a lot of code to be shared. I'd suggest you either:
 * have a helper function add pass appropriate functions for the trigger and the setup.
 * setup the environment and then call T.Run (https://pkg.go.dev/testing#T.Run)

Reply:
Ok updated both cgroupv1 and cgroupv2 with and without Exec context into their own helper functions.